kore/src/Kore/Step/SMT/Evaluator.hs

```diff
@@ -6,35 +6,121 @@ License     : NCSA
+                 ( when )
 import qualified Control.Monad.State.Strict as State
 import qualified Data.Map.Strict as Map
 import           Data.Reflection
 import qualified Data.Text as Text
 
 import qualified Control.Monad.Counter as Counter
-import           Kore.Predicate.Predicate
+import qualified Kore.Attribute.Symbol as Attribute
+                 ( Symbol )
+import           Kore.IndexedModule.MetadataTools
+                 ( SmtMetadataTools )
+import           Kore.Internal.Conditional
+                 ( Conditional )
+import qualified Kore.Internal.Conditional as Conditional
+import           Kore.Internal.MultiOr
+                 ( MultiOr )
+import qualified Kore.Internal.MultiOr as MultiOr
+import qualified Kore.Internal.Predicate as Predicate
+import qualified Kore.Predicate.Predicate as Syntax
                  ( Predicate )
+import qualified Kore.Predicate.Predicate as Syntax.Predicate
 import           Kore.Step.Simplification.Data
                  ( MonadSimplify )
 import qualified Kore.Step.Simplification.Data as Simplifier
 import           Kore.Step.SMT.Translate
                  ( Translator, evalTranslator, translatePredicate )
 import           Kore.Syntax.Variable
                  ( SortedVariable )
+import           Kore.TopBottom
+                 ( TopBottom )
 import           Kore.Unparser
-                 ( Unparse )
+                 ( Unparse, unparseToString )
 import           SMT
                  ( Result (..), SExpr (..) )
 import qualified SMT
 
+{- | Class for things that can be evaluated with an SMT solver,
+or which contain things that can be evaluated with an SMT solver.
+-}
+class Evaluable thing where
+    {- | Attempt to evaluate the argument wiith an external SMT solver.
+    -}
+    evaluate :: MonadSimplify m => thing -> m thing
+
+instance
+    ( SortedVariable variable
+    , Ord variable
+    , Show variable
+    , Unparse variable
+    )
+    => Evaluable (Syntax.Predicate variable)
+  where
+    evaluate predicate = do
+        refute <- decidePredicate predicate
+        return $ case refute of
+            Just False -> Syntax.Predicate.makeFalsePredicate
+            Just True -> Syntax.Predicate.makeTruePredicate
+            Nothing -> predicate
 
+instance
+    ( SortedVariable variable
+    , Ord variable
+    , Show variable
+    , Unparse variable
+    )
+    => Evaluable (Conditional variable term)
+  where
+    evaluate patt = do
+        evaluatedPredicate <- case Predicate.toPredicate predicate of
+            Syntax.Predicate.PredicateTrue -> return predicate
+            Syntax.Predicate.PredicateFalse -> return predicate
+            syntaxPredicate -> do
+                evaluatedPredicate <- evaluate syntaxPredicate
+                case evaluatedPredicate of
+                    Syntax.Predicate.PredicateTrue -> return Predicate.top
+                    Syntax.Predicate.PredicateFalse -> return Predicate.bottom
+                    _ -> do
```

Comment:
Well, the general idea is to apply whatever simplification the SMT can do, which, in principle, can go beyond a sat/unsat/always-sat answer. With that said, yes, that's the idea - convert to Top/Bottom whenever that's possible, or give up and return the predicate unchanged when we can't do that.

However, since we could, in principle, simplify the predicate in other ways, we have a problem: the predicate is a combination of the initial (syntax) predicate and the initial substitution, so we can't just return the simplified value without losing something. Possible fixes include: don't send the substitution, just apply it on the initial predicate before sending it to the SMT (probably the best option) and do a Haskell re-simplification of the SMT-simplified predicate, to extract the substitution. Actually, the first option may also contain potential substitutions in the SMT-simplified predicate, so we may also want to re-simplify it anyway. 

For the second issue:

First, a while ago we briefly considered sending both `p` and `not p` to the SMT solver, and simplifying both unsatisfiable predicates and always true predicates. I think that the conclusion was that we would probably use a lot of CPU time for that, and we wouldn't gain enough, but I'm not sure. Anyway, we didn't do that.

Second, returning a Bool from a `decidePredicate` function (or whatever it was called then) was more confusing than returning a `Maybe Bool`, which had a clear meaning (`Nothing` = we can't decide, `Just` = we know its truth value, and we're returning it). We could have used a custom type, but I'm not sure it was better.

So we ended up with a signature for a function which could give the predicate's truth value whenever that was possible, but which only returns `Nothing` and `Just False`, and which has a slim chance of returning `Just True` at some future point.

Reply:
> we could, in principle, simplify the predicate in other ways

The SMT simplification functions don't do that though, do they? (I think I'm wrong about that, maybe.) And if not, what is the returned pattern for?

Reply:
Hm, is it better now?